File: src/main/java/architecture/ego_equipment/datagen/tag/DatagenItemTag.java
package architecture.ego_equipment.datagen.tag;

import architecture.ego_equipment.core.EGOEquipment;
import architecture.ego_equipment.core.EGOEquipmentConstants;
import architecture.ego_equipment.init.tag.EquipmentItemTags;
import architecture.goldenboughs_lib.init.LibItems;
import architecture.goldenboughs_lib.init.tag.LibItemTags;
import net.minecraft.core.HolderLookup;
import net.minecraft.data.PackOutput;
import net.minecraft.data.tags.ItemTagsProvider;
import net.minecraft.tags.ItemTags;
import net.minecraft.tags.TagKey;
import net.minecraft.world.item.Item;
import net.minecraft.world.level.block.Block;
import net.neoforged.neoforge.common.Tags;
import net.neoforged.neoforge.common.data.ExistingFileHelper;
import net.neoforged.neoforge.registries.DeferredHolder;
import net.neoforged.neoforge.registries.DeferredItem;
import org.jetbrains.annotations.NotNull;

import java.util.Set;
import java.util.concurrent.CompletableFuture;

@SuppressWarnings("unchecked")
public final class DatagenItemTag extends ItemTagsProvider {
	public DatagenItemTag(final PackOutput output, final CompletableFuture<HolderLookup.Provider> lookupProvider, final CompletableFuture<TagLookup<Block>> blockTags, final ExistingFileHelper existingFileHelper) {
		super(output, lookupProvider, blockTags, EGOEquipment.ID, existingFileHelper);
	}

	@Override
	protected void addTags(final HolderLookup.Provider provider) {
		//region E.G.O
		addSet(EquipmentItemTags.EGO_ARMOUR, EGOEquipmentConstants.EGO_ARMOUR);
		addSet(EquipmentItemTags.EGO_TOOL, EGOEquipmentConstants.EGO_TOOL);
		addSet(EquipmentItemTags.EGO_WEAPON, EGOEquipmentConstants.EGO_WEAPON)
			.add(LibItems.CHAOS_SWORD.get());
		tag(LibItemTags.EGO)
			.addTags(EquipmentItemTags.EGO_ARMOUR,
				EquipmentItemTags.EGO_WEAPON,
				EquipmentItemTags.EGO_TOOL);

		addSet(ItemTags.HEAD_ARMOR, EGOEquipmentConstants.HEAD_ARMOR);
		addSet(ItemTags.CHEST_ARMOR, EGOEquipmentConstants.CHEST_ARMOR);
		addSet(ItemTags.LEG_ARMOR, EGOEquipmentConstants.LEG_ARMOR);
		addSet(ItemTags.FOOT_ARMOR, EGOEquipmentConstants.FOOT_ARMOR);
		tag(Tags.Items.ARMORS)
			.addTag(EquipmentItemTags.EGO_ARMOUR);
		//endregion

		//region 近战武器
		EGOEquipmentConstants.MELEE.addAll(EGOEquipmentConstants.KNIFE);
		EGOEquipmentConstants.MELEE.addAll(EGOEquipmentConstants.HAMMER);
		EGOEquipmentConstants.MELEE.addAll(EGOEquipmentConstants.FIST);
		EGOEquipmentConstants.MELEE.addAll(EGOEquipmentConstants.SPEAR);
		EGOEquipmentConstants.MELEE.addAll(EGOEquipmentConstants.MACE);
		addSet(EquipmentItemTags.MELEE, EGOEquipmentConstants.MELEE)
			.addTag(ItemTags.SWORDS);
		//endregion

		//region 远程武器
		EGOEquipmentConstants.GUN.addAll(EGOEquipmentConstants.CANNON);
		EGOEquipmentConstants.GUN.addAll(EGOEquipmentConstants.PISTOL);
		EGOEquipmentConstants.GUN.addAll(EGOEquipmentConstants.RIFLE);
		addSet(EquipmentItemTags.GUN, EGOEquipmentConstants.GUN);

		addSet(EquipmentItemTags.REMOTE, EGOEquipmentConstants.REMOTE).addTags(
			Tags.Items.TOOLS_CROSSBOW,
			Tags.Items.TOOLS_BOW,
			EquipmentItemTags.GUN);
		//endregion

		addSet(ItemTags.BOW_ENCHANTABLE, EGOEquipmentConstants.BOW);
		addSet(ItemTags.CROSSBOW_ENCHANTABLE, EGOEquipmentConstants.CROSSBOW);

		addSet(Tags.Items.TOOLS_CROSSBOW, EGOEquipmentConstants.CROSSBOW);
		addSet(Tags.Items.TOOLS_BOW, EGOEquipmentConstants.BOW);
		addSet(ItemTags.AXES, EGOEquipmentConstants.AXE);

		EGOEquipmentConstants.SWORDS.addAll(EGOEquipmentConstants.KNIFE);
		addSet(ItemTags.SWORDS, EGOEquipmentConstants.SWORDS)
			.add(LibItems.CHAOS_SWORD.get());
		tag(ItemTags.SWORD_ENCHANTABLE).addTag(EquipmentItemTags.MELEE);
		tag(ItemTags.BREAKS_DECORATED_POTS).addTag(EquipmentItemTags.MELEE);
		addSet(EquipmentItemTags.SPECIAL, EGOEquipmentConstants.SPECIAL);
		tag(Tags.Items.TOOLS)
			.add(LibItems.CREATIVE_RATIONALITY_TOOL.get())
			.addTags(
				EquipmentItemTags.EGO_WEAPON,
				EquipmentItemTags.EGO_TOOL,
				EquipmentItemTags.SPECIAL,
				EquipmentItemTags.REMOTE,
				EquipmentItemTags.MELEE);
	}

	private @NotNull IntrinsicTagAppender<Item> addSet(TagKey<Item> tag, @NotNull Set<DeferredItem<? extends Item>> set) {
		return tag(tag).add(set.stream().map(DeferredHolder::get).toArray(Item[]::new));
	}
}

File: src/main/kotlin/architecture/ego_equipment/datagen/tag/DatagenItemTag.kt
package architecture.ego_equipment.datagen.tag

import architecture.ego_equipment.init.tag.EgoEquipItemTags
import architecture.ego_equipment.util.EgoEquipUtil
import architecture.goldenboughs_lib.init.LibItems
import architecture.goldenboughs_lib.init.tag.LibItemTags
import net.minecraft.core.HolderLookup
import net.minecraft.data.PackOutput
import net.minecraft.data.tags.ItemTagsProvider
import net.minecraft.tags.ItemTags
import net.minecraft.tags.TagKey
import net.minecraft.world.item.Item
import net.minecraft.world.level.block.Block
import net.neoforged.neoforge.common.Tags
import net.neoforged.neoforge.common.data.ExistingFileHelper
import net.neoforged.neoforge.registries.DeferredItem
import java.util.concurrent.CompletableFuture

class DatagenItemTag(
	output: PackOutput,
	lookupProvider: CompletableFuture<HolderLookup.Provider>,
	blockTags: CompletableFuture<TagLookup<Block>>,
	existingFileHelper: ExistingFileHelper
) : ItemTagsProvider(output, lookupProvider, blockTags, EgoEquipUtil.ID, existingFileHelper) {

	override fun addTags(provider: HolderLookup.Provider) {
		addSet(EgoEquipItemTags.EGO_ARMOUR, EgoEquipUtil.EGO_ARMOUR)
		addSet(EgoEquipItemTags.EGO_TOOL, EgoEquipUtil.EGO_TOOL)
		addSet(EgoEquipItemTags.EGO_WEAPON, EgoEquipUtil.EGO_WEAPON)
			.add(LibItems.CHAOS_SWORD.get())
		tag(LibItemTags.EGO)
			.addTags(
				EgoEquipItemTags.EGO_ARMOUR,
				EgoEquipItemTags.EGO_WEAPON,
				EgoEquipItemTags.EGO_TOOL
			)

		addSet(ItemTags.HEAD_ARMOR, EgoEquipUtil.HEAD_ARMOR)
		addSet(ItemTags.CHEST_ARMOR, EgoEquipUtil.CHEST_ARMOR)
		addSet(ItemTags.LEG_ARMOR, EgoEquipUtil.LEG_ARMOR)
		addSet(ItemTags.FOOT_ARMOR, EgoEquipUtil.FOOT_ARMOR)
		tag(Tags.Items.ARMORS)
			.addTag(EgoEquipItemTags.EGO_ARMOUR)

		EgoEquipUtil.MELEE.addAll(EgoEquipUtil.KNIFE)
		EgoEquipUtil.MELEE.addAll(EgoEquipUtil.HAMMER)
		EgoEquipUtil.MELEE.addAll(EgoEquipUtil.FIST)
		EgoEquipUtil.MELEE.addAll(EgoEquipUtil.SPEAR)
		EgoEquipUtil.MELEE.addAll(EgoEquipUtil.MACE)
		addSet(EgoEquipItemTags.MELEE, EgoEquipUtil.MELEE)
			.addTag(ItemTags.SWORDS)

		EgoEquipUtil.GUN.addAll(EgoEquipUtil.CANNON)
		EgoEquipUtil.GUN.addAll(EgoEquipUtil.PISTOL)
		EgoEquipUtil.GUN.addAll(EgoEquipUtil.RIFLE)
		addSet(EgoEquipItemTags.GUN, EgoEquipUtil.GUN)

		addSet(EgoEquipItemTags.REMOTE, EgoEquipUtil.REMOTE).addTags(
			Tags.Items.TOOLS_CROSSBOW,
			Tags.Items.TOOLS_BOW,
			EgoEquipItemTags.GUN
		)

		addSet(ItemTags.BOW_ENCHANTABLE, EgoEquipUtil.BOW)
		addSet(ItemTags.CROSSBOW_ENCHANTABLE, EgoEquipUtil.CROSSBOW)

		addSet(Tags.Items.TOOLS_CROSSBOW, EgoEquipUtil.CROSSBOW)
		addSet(Tags.Items.TOOLS_BOW, EgoEquipUtil.BOW)
		addSet(ItemTags.AXES, EgoEquipUtil.AXE)

		EgoEquipUtil.SWORDS.addAll(EgoEquipUtil.KNIFE)
		addSet(ItemTags.SWORDS, EgoEquipUtil.SWORDS)
			.add(LibItems.CHAOS_SWORD.get())
		tag(ItemTags.SWORD_ENCHANTABLE).addTag(EgoEquipItemTags.MELEE)
		tag(ItemTags.BREAKS_DECORATED_POTS).addTag(EgoEquipItemTags.MELEE)
		addSet(EgoEquipItemTags.SPECIAL, EgoEquipUtil.SPECIAL)
		tag(Tags.Items.TOOLS)
			.add(LibItems.CREATIVE_RATIONALITY_TOOL.get())
			.addTags(
				EgoEquipItemTags.EGO_WEAPON,
				EgoEquipItemTags.EGO_TOOL,
				EgoEquipItemTags.SPECIAL,
				EgoEquipItemTags.REMOTE,
				EgoEquipItemTags.MELEE
			)
	}

	private fun addSet(tag: TagKey<Item>, set: MutableSet<DeferredItem<out Item>>) =
		tag(tag).add(*set.map { it.get() }.toTypedArray())
}
